docs: add how_to_build (#322)
kroma-network/tachyon: docs/how_to_use/how_to_build.md

  ```shell
  bazel build --config ${os} --config avx512_${os} --//:polygon_zkevm_backend //...
  ```

### Hardware acceleration

#### CUDA backend

- `--config cuda`: Enable [cuda] backend.

  ```shell
  bazel build --config ${os} --config cuda //...
  ```

#### ROCm backend

- `--config rocm`: Enable [rocm] backend.

  ```shell
  bazel build --config ${os} --config rocm //...
  ```

_NOTE_: The `rocm` option is not recommended for current use because it is not being tested yet.

[cuda]: https://developer.nvidia.com/cuda-toolkit
[rocm]: https://www.amd.com/en/graphics/servers-solutions-rocm

### .bazelrc.user

Build options can be preset in `.bazelrc.user` for your convenience, eliminating the need to specify them on the command line.

For example:

```
# .bazelrc.user

build --config linux
build --config dbg
build --config polygon_zkevm_backend
```

```shell
bazel build //...
# With the preset options in .bazelrc.user, this is the same as:
# bazel build --config linux --config dbg --config polygon_zkevm_backend //...
```

## Building Tachyon from a Bazel repository

> Tachyon can be built in your own Bazel project with the following two simple steps.
> First, obtain the Tachyon code from a specific commit hash and get a SHA256 value from the fetched code through the these commands:

Reviewer: ```suggestion
Tachyon can be built in your own Bazel project with the following two simple steps:

First, obtain the Tachyon code from a specific commit hash and get a SHA256 value from the fetched code through these commands:
```
mb

Reviewer: Change "through the these commands:" to  "through these commands:"

[cuda]: https://developer.nvidia.com/cuda-toolkit
[rocm]: https://www.amd.com/en/graphics/servers-solutions-rocm

### .bazelrc.user

You can set the options in `.bazelrc.user`.

For example:

```
build --config linux
build --config dbg
build --config polygon_zkevm_backend
```

If you add `.bazelrc.user` like above, you can build with the options without specify in the command.

```shell
bazel build //...
# It is same as:
# bazel build --config linux --config dbg --config polygon_zkevm_backend //...
```

## Build from Bazel repository

You can build Tachyon on your own bazel project. The following snippet is an example of `WORKSPACE` to use Tachyon in an external project.
You can get the Tachyon repo in specific commit. You'll also need a sha256 value of the fetched code.

```shell
wget https://github.com/kroma-network/tachyon/archive/c6a38895cb6f9b9f73e3d64928d85a9762b66639.tar.gz

shasum -a 256 c6a38895cb6f9b9f73e3d64928d85a9762b66639.tar.gz
3ebc4edca153c9b93de200d5e22be66d1ff7ea52e2793c83947607c532dfacbc  c6a38895cb6f9b9f73e3d64928d85a9762b66639.tar.gz
```

> ```py

Reviewer 2: Please use `bzl`. See https://raw.githubusercontent.com/bazelbuild/buildtools/master/buildifier/README.md.

> // WORKSPACE 

Reviewer 2: ```suggestion
# WORKSPACE
```

load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
http_archive(
    name = "kroma_network_tachyon",
    sha256 = "3ebc4edca153c9b93de200d5e22be66d1ff7ea52e2793c83947607c532dfacbc",
    strip_prefix = "tachyon-c6a38895cb6f9b9f73e3d64928d85a9762b66639",
    urls = ["https://github.com/kroma-network/tachyon/archive/c6a38895cb6f9b9f73e3d64928d85a9762b66639.tar.gz"],
)
```

## Debian packaging

Build a Debian package with the supported scheme(only halo2 for now) and the [options](#options) you want.

> For example, to build the Halo2 Debian package with `polygon_zkevm_backend` and `has_openmp` options, run the following command:

Reviewer 2: You should guide that this is the recommended flag.

```shell
bazel build -c opt --config halo2 --//:polygon_zkevm_backend --//:has_openmp  --//:c_shared_object //scripts/packages/debian/runtime:debian
bazel build -c opt --config halo2 --//:polygon_zkevm_backend --//:has_openmp  --//:c_shared_object //scripts/packages/debian/dev:debian

sudo dpkg -i bazel-bin/scripts/packages/debian/runtime/libtachyon_0.0.1_amd64.deb
sudo dpkg -i bazel-bin/scripts/packages/debian/dev/libtachyon-dev_0.0.1_amd64.deb

```

## Frequently Asked Questions

> #### Debugging on macOS

Reviewer 2: Did you turn on markdown lint? you have to increase the heading level one by one.

  CONFIGURE_OPTS=--enable-shared pyenv install <version>
  ```

#### Matplotlib(optional)

If you want to print a benchmark plot, you need to install matplotlib. See [here](/benchmark/README.md) for more details.

```shell
pip install matplotlib
```

**_NOTE_:**

- To build `mac_logging` in MacOS, you need to install objc, which can be simply done by installing XCode.

- MacOS v14.0.0 or later is recommended. In certain versions of MacOS (prior to v13.5.1), a bug related to incorrect `BigInt` divide operations has been detected in the field generator when using the optimized build (`-c opt`). This [issue](https://github.com/kroma-network/tachyon/issues/98) will be fixed as soon as possible.

## Build from source

### Build

#### Build on Linux

```shell
bazel build --config linux //...
```

#### Build on Macos arm64

```shell
bazel build --config macos_arm64 //...
```

#### Build on Macos x64

```shell
bazel build --config macos_x86_64 //...
```

### Test

#### Test on Linux

```shell
bazel test --config linux //...
```

#### Test on Macos arm64

```shell
bazel test --config macos_arm64 //...
```

#### Test on Macos x64

```shell
bazel test --config macos_x86_64 //...
```

> ## Options

Author: Makes sense. What about "Configurations"?

> ## Options

Reviewer: Sg~

### Build options

- `opt`: Default optimized build option
- `dbg`: Build with debug info
- `fastbuild`: Fast build option

```shell
bazel build --config macos_arm64 --config dbg //...
```

### Polygon zkEVM backend prime field

_NOTE_: Only x86_64 is supported.

- `--//:polygon_zkevm_backend`: Enable [goldilocks](https://github.com/0xPolygonHermez/goldilocks) and [zkevm-prover](https://github.com/0xPolygonHermez/zkevm-prover) prime field backend.

  ```shell
  bazel build --config ${os} --config avx512_${os} --//:polygon_zkevm_backend //...
  ```

### Hardware acceleration

#### CUDA backend

- `--config cuda`: Enable [cuda] backend.

  ```shell
  bazel build --config ${os} --config cuda //...
  ```

#### ROCm backend

- `--config rocm`: Enable [rocm] backend.

  ```shell
  bazel build --config ${os} --config rocm //...
  ```

_NOTE_: The `rocm` option is not recommended for current use because it is not being tested yet.

[cuda]: https://developer.nvidia.com/cuda-toolkit
[rocm]: https://www.amd.com/en/graphics/servers-solutions-rocm

### .bazelrc.user

You can preset the options in `.bazelrc.user`.

For example:

> ```

Reviewer: or is it # instead of //?

build --config linux
build --config dbg
build --config polygon_zkevm_backend
```

By including options in .bazelrc.user, you can build with those options automatically, eliminating the need to specify them on the command line.

```shell
bazel build //...

> # This is same as:

Reviewer: ``` suggestion
# With the preset options in .bazelrc.user, this is the same as:
```

# bazel build --config linux --config dbg --config polygon_zkevm_backend //...
```

> ## Build from Bazel repository
> 
> You can build Tachyon in your own Bazel project. The following snippet is an example of a `WORKSPACE` that uses Tachyon in an external project.
> 
> For the first step, obtain the Tachyon code from a specific commit hash and get a SHA256 value from the fetched code.
> 
> ```shell
> wget https://github.com/kroma-network/tachyon/archive/c6a38895cb6f9b9f73e3d64928d85a9762b66639.tar.gz
> 
> shasum -a 256 c6a38895cb6f9b9f73e3d64928d85a9762b66639.tar.gz
> ```
> 
> After receiving a shasum output such as `3ebc4edca153c9b93de200d5e22be66d1ff7ea52e2793c83947607c532dfacbc`, input it into `WORKSPACE` like the following:
> 
> ```bzl
> # WORKSPACE 
> 
> load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
> http_archive(
>     name = "kroma_network_tachyon",
>     sha256 = "3ebc4edca153c9b93de200d5e22be66d1ff7ea52e2793c83947607c532dfacbc",
>     strip_prefix = "tachyon-c6a38895cb6f9b9f73e3d64928d85a9762b66639",
>     urls = ["https://github.com/kroma-network/tachyon/archive/c6a38895cb6f9b9f73e3d64928d85a9762b66639.tar.gz"],
> )
> ```

Reviewer: ```suggestion
## Building Tachyon from a Bazel repository

Tachyon can be built in your own Bazel project with the following two simple steps.

First, obtain the Tachyon code from a specific commit hash and get a SHA256 value from the fetched code through the these commands:

 ```shell
 wget https://github.com/kroma-network/tachyon/archive/c6a38895cb6f9b9f73e3d64928d85a9762b66639.tar.gz

 shasum -a 256 c6a38895cb6f9b9f73e3d64928d85a9762b66639.tar.gz
 ```

 Second, input the shasum output into your `WORKSPACE` file as the `sha256` argument like shown below:

 ```bzl
 # WORKSPACE 
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 http_archive(
     name = "kroma_network_tachyon",
     sha256 = "3ebc4edca153c9b93de200d5e22be66d1ff7ea52e2793c83947607c532dfacbc",
     strip_prefix = "tachyon-c6a38895cb6f9b9f73e3d64928d85a9762b66639",
     urls = ["https://github.com/kroma-network/tachyon/archive/c6a38895cb6f9b9f73e3d64928d85a9762b66639.tar.gz"],
 )
 ```
```
```

the suggestion format got a bit mixed up, but hopefully you can tell what I mean.

sudo dpkg -i bazel-bin/scripts/packages/debian/runtime/libtachyon_0.0.1_amd64.deb
sudo dpkg -i bazel-bin/scripts/packages/debian/dev/libtachyon-dev_0.0.1_amd64.deb

```

## Other Info

### Debugging on macOS

Please add this line to your `.bazelrc.user`.

```
build --spawn_strategy=local
```

### Py Binding

`ModuleNotFoundError` may occur in certain python versions (v3.11.6). Python v3.10.12 is recommended.

### Build on Ubuntu 20.04

If you are using Ubuntu 20.04, update your g++ version. The default `g++-9` does not work.

```shell
sudo apt install g++-10
export CC=/usr/bin/gcc-10
export CXX=/usr/bin/g++-10
export GCC_HOST_COMPILER_PATH=/usr/bin/gcc-10
```

### Build CUDA with rust toolchain

You may run into the following problem:

```shell
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
error: linking with `external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc` failed: exit status: 127
...
  = note: /usr/bin/env: 'python': No such file or directory
```

If so, make your `python` point to the python interpreter.

```shell
sudo apt install python-is-python3
```

> Additionally, please include these lines to your `.bazelc.user`.

Reviewer: ```suggestion
Additionally, please include these lines in your `.bazelc.user`.
